### PR TITLE
Add examples

### PR DESCRIPTION
**Pull request by ebezzam**
Will be useful for paper. I've also added some convolution examples and functions, I know this is a point to discuss as @SepandKashani wanted to keep it separate, but I think it might be useful to show such examples in the package so people have something to get inspiration from. Maybe the convolution functions can be moved into the example script rather than in the package?

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -23,6 +23,7 @@ classifiers =
     Programming Language :: Python :: Implementation :: CPython
     Topic :: Scientific/Engineering
 license = BSD
+version = 2.1.1
```
Comment (SepandKashani) [inline setup.cfg:26]: This is inferred automatically from the git commits. No need to include it.

File: pyffs/util.py [hunk 2]
```diff
@@ -425,3 +426,34 @@ def _create_modulation_vectors(N_s, N_FS, T, T_c, mod=None):
         E_2 = mod.r_[mod.arange(start=0, stop=M), mod.arange(start=-M, stop=0)]
 
     return B_1 ** E_1, B_2 ** (N * E_2)
+
+
+def ffsn_shift(x, idx):
```
Comment (SepandKashani) [inline pyffs/util.py:431]: Don't understand what this does?

File: profile/util.py [hunk 3]
```diff
@@ -1,5 +1,22 @@
...
+
+def plotting_setup(font_size=30, linewidth=4, markersize=10, fig_folder="figs"):
+    font = {"family": "Times New Roman", "weight": "normal", "size": font_size}
+    matplotlib.rc("font", **font)
+    matplotlib.rcParams["lines.linewidth"] = linewidth
+    matplotlib.rcParams["lines.markersize"] = markersize
+
+    fig_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), fig_folder)
+    if not os.path.isdir(fig_path):
+        os.mkdir(fig_path)
```
Comment (SepandKashani) [inline profile/util.py:18]: Use `pathlib.Path` instead:
```
import pathlib as plib
fig_path = plib.Path(__file__).parent / fig_folder
fig_path.mkdir(exist_ok=True)
```

File: profile/ffsn.py [hunk 4]
```diff
@@ -62,12 +73,14 @@ def profile_ffsn(n_trials):
...
+    comparison_plot(proc_time, proc_time_std, n_std, ax)
     ax.set_xlabel("Number of FS coefficients")
+    ax.set_xticks(np.array(N_FS_vals) - 1)
+    ax.set_yticks([1e-4, 1e-3, 1e-2, 1e-1, 1e0, 1e1])
     fig.tight_layout()
+    fig.savefig(os.path.join(fig_path, "profile_ffsn_2d.png"))
 
-    fname = pathlib.Path(__file__).resolve().parent / "profile_ffsn_2d.png"
-    fig.savefig(fname, dpi=300)
+    plt.show()
```
Comment (SepandKashani) [inline profile/ffsn.py:83]: `fig.show()`

File: profile/ffs.py [hunk 5]
```diff
@@ -55,12 +58,14 @@ def profile_ffsn(n_trials):
...
+    comparison_plot(proc_time, proc_time_std, n_std, ax)
     ax.set_xlabel("Number of FS coefficients")
+    ax.set_xticks(np.array(N_FS_vals) - 1)
+    ax.set_yticks([1e-4, 1e-3, 1e-2, 1e-1])
     fig.tight_layout()
+    fig.savefig(os.path.join(fig_path, "profile_ffs.png"))
 
-    fname = pathlib.Path(__file__).resolve().parent / "profile_ffs.png"
-    fig.savefig(fname, dpi=300)
+    plt.show()
```
Comment (SepandKashani) [inline profile/ffs.py:68]: `fig.show()`

File: profile/convolve2d.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    # plot results
+    fig, ax = plt.subplots()
+    comparison_plot(proc_time, proc_time_std, n_std, ax)
+    ax.set_xlabel("Number of samples per dimension")
+    ax.set_xticks([10, 30, 100, 300])
+    fig.tight_layout()
+    fig.savefig(os.path.join(fig_path, "profile_convolve2d.png"))
+
+    plt.show()
```
Comment (SepandKashani) [inline profile/convolve2d.py:71]: `fig.show()`

File: profile/bandlimited_interp2d_vary_width.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    # plot results
+    fig, ax = plt.subplots()
+    comparison_plot(proc_time, proc_time_std, n_std, ax)
+    ax.set_title(f"{N_s} samples, {M} interp points")
+    ax.set_xlabel("Percentage of period")
+    fig.tight_layout()
+    fig.savefig(os.path.join(fig_path, "bandlimited_interp2d_vary_width.png"))
+
+    plt.show()
```
Comment (SepandKashani) [inline profile/bandlimited_interp2d_vary_width.py:98]: `fig.show()`

File: profile/bandlimited_interp2d_vary_M.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+    # plot results
+    fig, ax = plt.subplots()
+    comparison_plot(proc_time, proc_time_std, n_std, ax)
+    ax.set_title(f"{N_s} samples, {percent_region*100}% of period")
+    ax.set_xlabel("Number of interpolation points in section")
+    fig.tight_layout()
+    fig.savefig(os.path.join(fig_path, "bandlimited_interp2d_vary_M.png"))
+
+    plt.show()
```
Comment (SepandKashani) [inline profile/bandlimited_interp2d_vary_M.py:121]: `fig.show()`

File: profile/bandlimited_interp1d_vary_width.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    # plot results
+    fig, ax = plt.subplots()
+    comparison_plot(proc_time, proc_time_std, n_std, ax)
+    ax.legend(loc="upper right")
+    ax.set_title(f"{n_samples} samples, {M} interpolation points")
+    ax.set_xlabel("Percentage of period")
+    fig.tight_layout()
+    fig.savefig(os.path.join(fig_path, "bandlimited_interp1d_vary_width.png"))
+
+    plt.show()
```
Comment (SepandKashani) [inline profile/bandlimited_interp1d_vary_width.py:84]: `fig.show()`

File: profile/bandlimited_interp1d_vary_M.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+    # # plot theoretical crossover
+    # m = (stop - start) / T
+    # dt = T / N_FS * (1 - m)
+    # M_cross = int(m * T / dt)
+    # if M_cross > 0:
+    #     ax.axvline(x=M_cross, linestyle="--", color="g")
+    fig.savefig(os.path.join(fig_path, "bandlimited_interp1d_vary_M.png"))
+
+    plt.show()
```
Comment (SepandKashani) [inline profile/bandlimited_interp1d_vary_M.py:133]: `fig.show()`

File: examples/ffs_1d.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+T, T_c, N_FS = math.pi, math.e, 15
+N_samples = 512
+sample_points, _ = ffs_sample(T, N_FS, T_c, N_samples, mod=np)
+diric_samples = dirichlet(sample_points, T, T_c, N_FS)
+diric_FS = ffs(diric_samples, T, T_c, N_FS)[:N_FS]
+
+# plot time
+idx = np.argsort(sample_points)
+_, ax = plt.subplots(nrows=2, num="FS coefficients of Dirichlet", figsize=(10, 10))
```
Comment (SepandKashani) [inline examples/ffs_1d.py:19]: `fig, ax = plt.subplots(nrows=2, num="FS coefficients of Dirichlet", figsize=(10, 10))`

File: examples/ffs_1d.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+# plot frequency
+N = N_FS // 2
+fs_idx = np.arange(-N, N + 1)
+ax[1].stem(fs_idx, np.abs(diric_FS))
+ax[1].grid()
+ax[1].set_xlabel("FS index")
+
+plt.savefig(os.path.join(fig_path, "ffs_1d.png"))
+plt.show()
```
Comment (SepandKashani) [inline examples/ffs_1d.py:33]:
```
fig.savefig(plib.Path(fig_path) / "ffs_1d.png")
fig.show()
```

File: examples/ffs_2d.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+ky_vals = np.arange(-K_plot[1], K_plot[1] + 1)
+X, Y = np.meshgrid(kx_vals, ky_vals)
+cp = ax_freq.contourf(X, Y, np.abs(FS.T))
+fig = plt.gcf()
+fig.colorbar(cp)
+ax_freq.set_xlabel("$FS_x$")
+ax_freq.set_ylabel("$FS_y$")
+
+plt.savefig(os.path.join(fig_path, "ffs_2d.png"))
+plt.show()
```
Comment (SepandKashani) [inline examples/ffs_2d.py:49]: fig.savefig / fig.show

File: examples/interp_1d.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    x=start,
+    c="k",
+    linestyle="--",
+)
+ax.axvline(
+    x=stop,
+    c="k",
+    linestyle="--",
+)
+plt.legend()
```
Comment (SepandKashani) [inline examples/interp_1d.py:73]: ax.legend/tigth_layout/savefig()

File: examples/interp_1d.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+# ax.plot(points, vals_sinc, label="sinc interp")
+ax.plot(resampled_t, resampled_x, label="scipy.signal.resample", alpha=ALPHA, linestyle="-.")
+ax.scatter(sample_points, diric_samples, label="available samples")
+ax.set_xlabel("Time [s]")
+ax.set_xlim([start, stop])
+plt.legend()
+plt.tight_layout()
+plt.savefig(os.path.join(fig_path, "interp_1d.png"))
+
+plt.show()
```
Comment (SepandKashani) [inline examples/interp_1d.py:102]: fig vs plt

File: examples/util.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        y_vals = np.append(y_vals, [y_vals[-1] + dy])
+
+    fig = plt.figure()
+    ax = fig.add_subplot(1, 1, 1)
+    X, Y = np.meshgrid(x_vals, y_vals)
+    if pcolormesh:
+        cp = ax.pcolormesh(X, Y, Z.T)
+    else:
+        cp = ax.contourf(X, Y, Z.T)
+    fig = plt.gcf()
```
Comment (SepandKashani) [inline examples/util.py:56]: You have a reference to `fig` on L49. Why is this required?

File: examples/convolve_1d.py [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    alpha=ALPHA,
+)
+ax[1].set_xlim([np.min(sample_points), np.max(sample_points)])
+ax[1].set_ylabel("$f \\ast f$")
+ax[1].set_xlabel("Time [s]")
+plt.legend()
+plt.tight_layout()
+plt.savefig(os.path.join(fig_path, "convolve_1d.png"))
+
+plt.show()
```
Comment (SepandKashani) [inline examples/convolve_1d.py:70]: fig vs plt.

File: examples/convolve_2d.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
+import numpy as np
```
Comment (SepandKashani) [inline examples/convolve_2d.py:1]: fig vs plt

**Review by SepandKashani — COMMENTED** [review]
Minor suggestions.
I have some doubts on the API of `pyffs/conv.py`, but let's go with this for now.

File: pyffs/util.py (hunk 2 above)
Comment (ebezzam) [reply pyffs/util.py:431]: It's like `fftshift`: https://numpy.org/doc/stable/reference/generated/numpy.fft.fftshift.html
For higher dimensions, it gets clunky to re-order the samples with the output of `ffsn_sample`. So I thought a utility function could allow for more compact code. Could also provide the same args as those provided to `ffsn_sample` so that the indices are computed internally, but then that's more arguments...

File: profile/ffsn.py (hunk 4 above)
Comment (ebezzam) [reply profile/ffsn.py:83]: Perhaps it's my matplotlib settings, but `fig.show()` doesn't generate a pop-up window with the figure, while `plt.show()` does.

**Review by SepandKashani — APPROVED** [review]
